File: java/java-samples/src/main/java/tech/techstay/oop/Interface.java
package tech.techstay.oop;

interface MyInterface {
  // static methods
  static void baz() {
    System.out.println("baz");
  }

  // abstract methods
  void foo();

  // default methods
  default void bar() {
    System.out.println("bar");
  }
}


public class Interface {
  public static void main(String[] args) {
    MyInterface my = new MyClass();
    my.foo();
    my.bar();
    MyInterface.baz();
  }
}


class MyClass implements MyInterface {
  @Override
  public void foo() {
    System.out.println("foo");
  }
}

File: kotlin/kotlin-demo/src/main/kotlin/tech/techstay/oop/Interface.kt
package tech.techstay.oop

interface MyInterface {
    fun foo()
    fun bar(name: String): String
}

class MyImplements : MyInterface {
    override fun foo() {
        println("implemented foo")
    }

    override fun bar(name: String): String {
        println("implemented bar")
        return name.uppercase()
    }

}

fun main() {
    val obj = MyImplements()
    obj.foo()
    println(obj.bar("techstay"))
}
